mbw.c

```diff
@@ -99,40 +138,44 @@ long *make_array(unsigned long long asize)
  *
  * return value: elapsed time in seconds
  */
-double worker(unsigned long long asize, long *a, long *b, int type, unsigned long long block_size)
+double worker(unsigned long long asize, long *a, long *b, int type, unsigned long long block_size, int num_workers)
 {
-    unsigned long long t;
     struct timeval starttime, endtime;
     double te;
-    unsigned int long_size=sizeof(long);
-    /* array size in bytes */
-    unsigned long long array_bytes=asize*long_size;
-
-    if(type==TEST_MEMCPY) { /* memcpy test */
-        /* timer starts */
-        gettimeofday(&starttime, NULL);
-        memcpy(b, a, array_bytes);
-        /* timer stops */
-        gettimeofday(&endtime, NULL);
-    } else if(type==TEST_MCBLOCK) { /* memcpy block test */
-        char* src = (char*)a;
-        char* dst = (char*)b;
-        gettimeofday(&starttime, NULL);
-        for (t=array_bytes; t >= block_size; t-=block_size, src+=block_size){
-            dst=(char *) memcpy(dst, src, block_size) + block_size;
-        }
-        if(t) {
-            dst=(char *) memcpy(dst, src, t) + t;
-        }
-        gettimeofday(&endtime, NULL);
-    } else if(type==TEST_DUMB) { /* dumb test */
-        gettimeofday(&starttime, NULL);
-        for(t=0; t<asize; t++) {
-            b[t]=a[t];
-        }
-        gettimeofday(&endtime, NULL);
+    int i;
+
+    pthread_t *threads = malloc(num_workers * sizeof(pthread_t));
+    thread_param_t *params = malloc(num_workers * sizeof(thread_param_t));
+    unsigned long long chunk_asize = asize / num_workers;
```

Comment:
suggestion: it would be more intuitive to keep asize (and thereby use a total of num_workers*asize memory), just need to document this. It would also sidestep the rounding math.